pageserver/src/tenant/timeline.rs

```diff
@@ -993,6 +1013,15 @@ impl Timeline {
         // now lock out layer removal (compaction, gc, timeline deletion)
         let layer_removal_guard = self.layer_removal_cs.lock().await;
 
+        {
+            // to avoid racing with detach and delete_timeline
```

Comment:
Avoid != prevent. IIRC this is mostly to avoid misleading `error!` messages.

For timeline delete, the `layer_removal_cs` + this snippet help to avoid:
```
B: collects list of layers to evict, observes timeline as active
A: timeline_delete sets timeline to stopping state
B: calls this function
A: timeline_delete removes layer files from disk (they are still referenced in layermap though)
B: successfully removes the evictees from layer map
B: fails to remove the file from disk because it's already deleted => logs an error!
```

For tenant detach, you are right, this doesn't help. Perhaps tenant detach should also grab `layer_removal_cs`?